Add `nvidia_smi` Plugin to help monitor GPU heavy rigs

jackzampolin opened the pull request:
This PR implements a simple version of a GPU monitoring plugin. 

This PR closes https://github.com/influxdata/telegraf/issues/2642 

### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

danielnelson commented:
Hey Jack, I'll try to take a look later today, but does this command usually require root permissions?

jackzampolin commented:
@danielnelson 👋its not looking like it on the `ethos` system I'm using to test this. I can't speak to other systems.

danielnelson commented on plugins/inputs/nvidia_smi/nvidia_smi.go:
```diff
@@ -0,0 +1,127 @@
+	}
+
+	retS := strings.TrimSuffix(string(ret), "\n")
+	retI, errI := strconv.Atoi(retS)
```
Use `strconv.ParseInt`, this avoids values being parsed as 32-bit ints on 32-bit systems.

danielnelson commented on plugins/inputs/nvidia_smi/nvidia_smi.go:
```diff
@@ -0,0 +1,127 @@
+	fields := make(map[string]interface{}, 0)
+
+	// Construct and execute metrics query
+	opts := []string{"--format=noheader,nounits,csv", fmt.Sprintf("--query-gpu=%s", smi.metrics), fmt.Sprintf("--id=%d", gpuID)}
```
Check if nvidia-smi supports space delimited options so we can avoid the Sprintf: `--id <gpuID>` instead of `--id=<gpuID>`.

jackzampolin replied:
Does not support space delimited options 😢

danielnelson commented on plugins/inputs/nvidia_smi/nvidia_smi.go:
```diff
@@ -0,0 +1,127 @@
+
+	// Construct and execute metrics query
+	opts := []string{"--format=noheader,nounits,csv", fmt.Sprintf("--query-gpu=%s", smi.metrics), fmt.Sprintf("--id=%d", gpuID)}
+	ret, err := exec.Command(smi.BinPath, opts...).CombinedOutput()
```
`internal.CombinedOutputTimeout`

danielnelson commented on plugins/inputs/nvidia_smi/README.md:
```diff
@@ -0,0 +1,25 @@
+# `nvidia-smi` Input Plugin
```
Can you add the missing sections and update style from: https://github.com/influxdata/telegraf/blob/master/plugins/inputs/EXAMPLE_README.md

jackzampolin replied:
🙊 I'll add these!

jackzampolin replied:
This is done!

danielnelson commented on plugins/inputs/nvidia_smi/nvidia_smi.go:
```diff
@@ -0,0 +1,127 @@
+	met := strings.Split(string(ret), ", ")
+	for i, m := range metricNames {
+		if m[1] == "tag" {
+			tags[m[0]] = strings.TrimSuffix(met[i], "\n")
```
Must guard against out of range for `met[i]`, probably before the loop check `len(met) == len(metricNames)`

danielnelson replied:
`strings.TrimSpace`

danielnelson commented on plugins/inputs/nvidia_smi/README.md:
```diff
@@ -0,0 +1,25 @@
+- `compute_mode`: The compute mode of the GPU e.g. `Default`
+- `index`: The port index where the GPU is connected to the motherboard e.g. `1`
+- `pstate`: Overclocking state for the GPU e.g. `P0`
+- `uuid`: A unique identifier for the GPU e.g. `GPU-f9ba66fc-a7f5-94c5-da19-019ef2f9c665`
```
Do you know when this changes?  Can you check on reboot and, if possible, on change of slot?

jackzampolin replied:
Good question! On reboot it does not change. I'll have to try change of slot next time I have access to the machines.

jackzampolin commented:
The queries without the `--id={{foo}}` in there. Working on refactoring this now. How would you suggest coding unit tests? Similar to the [`nstat`](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/nstat/nstat_test.go) plugin?

jackzampolin commented:
Alright @danielnelson I've addressed all the PR comments, updated the README and added tests. Sorry for not having this in good shape on the first push 🙊

danielnelson commented on plugins/inputs/nvidia_smi/nvidia_smi.go:
```diff
@@ -0,0 +1,143 @@
+
+func gatherNvidiaSMI(ret string, acc telegraf.Accumulator) error {
+	// First split the lines up and handle each one
+	lines := strings.Split(string(ret), "\n")
```
Can you update this to also work with DOS line endings?  I usually just use `bufio.Scanner`.

jackzampolin replied:
Done!

danielnelson commented on plugins/inputs/nvidia_smi/nvidia_smi.go:
```diff
@@ -0,0 +1,143 @@
+	met := strings.Split(line, ",")
+
+	// Make sure there are as many metrics in the line as there were queried.
+	if len(met) == len(metricNames) {
```
I think this should return an error when they are not equal.

jackzampolin replied:
Done!

jackzampolin commented:
Did those last nits!

danielnelson commented:
Thanks! :beers:

timhallinflux commented:
@dp1140a -- FYI

deric commented:
@jackzampolin string timeout, like mentioned in the example:
```
[[inputs.nvidia_smi]]
  timeout = "5s"
```
is currently not parsed correctly:
```
2018/04/24 09:24:01 E! Error parsing /etc/telegraf/telegraf.conf, line 61: nvidia_smi.NvidiaSMI.Timeout: `string' type is not assignable to `time.Duration' type
```

jackzampolin commented:
@deric Argh. Sorry about that. I've got a PR up to fix: https://github.com/influxdata/telegraf/pull/4069

gsxryan commented:
I stumbled here to look for the addition of Watts measurement - power.draw .  This has been added in #4262: Add power draw field to nvidia_smi plugin.
